src/Map.ts

```diff
@@ -0,0 +1,825 @@
+  const lookupS = lookup(S)
+  return (k, m) => lookupS(k, m).isSome()
+}
+
+/**
+ * Test whether or not a value is a member of a map
+ *
+ * @since 1.14.0
+ */
+export const isMember = <A>(S: Setoid<A>): (<K>(a: A, m: Map<K, A>) => boolean) => {
+  return (a, m) => {
+    const as = m.values()
+    let e: IteratorResult<A>
+    while (!(e = as.next()).done) {
+      if (S.equals(a, e.value)) {
+        return true
+      }
+    }
+    return false
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const keys = <K>(O: Ord<K>): (<A>(m: Map<K, A>) => Array<K>) => {
+  const sortO = sort(O)
+  return m => {
+    const ks = Array.from(m.keys())
+    return sortO(ks)
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const keysSet = <K, A>(m: Map<K, A>): Set<K> => new Set(m.keys())
+
+/**
+ * @since 1.14.0
+ */
+export const collect = <K>(O: Ord<K>): (<A, B>(m: Map<K, A>, f: (k: K, a: A) => B) => Array<B>) => {
+  const lookupWithKeyO = lookupWithKey(O)
+  const keysO = keys(O)
+  return <A, B>(m: Map<K, A>, f: (k: K, a: A) => B): Array<B> => {
+    const out: Array<B> = []
+    const ks = keysO(m)
+    for (let key of ks) {
+      const o = lookupWithKeyO(key, m)
+      if (o.isSome()) {
+        out.push(f(key, o.value[1]))
+      }
+    }
+    return out
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const toArray = <K>(O: Ord<K>): (<A>(m: Map<K, A>) => Array<[K, A]>) => {
+  const collectO = collect(O)
+  return <A>(m: Map<K, A>): Array<[K, A]> => collectO(m, (k, a: A) => tuple(k, a))
+}
+
+/**
+ * Unfolds a Map into a list of key/value pairs
+ *
+ * @since 1.14.0
+ */
+export function toUnfoldable<K, F extends URIS>(
+  O: Ord<K>,
+  unfoldable: Unfoldable1<F>
+): <A>(d: Map<K, A>) => Type<F, [K, A]>
+export function toUnfoldable<K, F>(O: Ord<K>, unfoldable: Unfoldable<F>): <A>(d: Map<K, A>) => HKT<F, [K, A]>
+export function toUnfoldable<K, F>(O: Ord<K>, unfoldable: Unfoldable<F>): <A>(d: Map<K, A>) => HKT<F, [K, A]> {
+  const toArrayO = toArray(O)
+  return d => {
+    const arr = toArrayO(d)
+    const len = arr.length
+    return unfoldable.unfoldr(0, b => (b < len ? some(tuple(arr[b], b + 1)) : none))
+  }
+}
+
+/**
+ * Insert or replace a key/value pair in a Map
+ *
+ * @since 1.14.0
+ */
+export const insert = <K>(S: Setoid<K>): (<A>(k: K, a: A, m: Map<K, A>) => Map<K, A>) => {
+  const lookupS = lookupWithKey(S)
+  return (k, a, m) => {
+    const found = lookupS(k, m)
+    if (found.isNone()) {
+      const r = new Map(m)
+      r.set(k, a)
+      return r
+    } else if (found.value[1] !== a) {
+      const r = new Map(m)
+      r.set(found.value[0], a)
+      return r
+    }
+    return m
+  }
+}
+
+/**
+ * Delete a key and value from a map
+ *
+ * @since 1.14.0
+ */
+export const remove = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Map<K, A>) => {
+  const lookupS = lookupWithKey(S)
+  return (k, m) => {
+    const found = lookupS(k, m)
+    if (found.isSome()) {
+      const r = new Map(m)
+      m.delete(k)
```

Comment:
shouldn't change `m` and should use the existing key

```diff
-      m.delete(k)
+      r.delete(found.value[0])
```

Reply:
I made this change. Marking it as resolved.